deps/rabbit/src/rabbit_fifo_client.erl

```diff
@@ -320,29 +328,45 @@ discard(ConsumerTag, [_|_] = MsgIds,
 %%
 %% @returns `{ok, State}' or `{error | timeout, term()}'
 -spec checkout(rabbit_types:ctag(),
-               NumUnsettled :: non_neg_integer(),
                CreditMode :: rabbit_fifo:credit_mode(),
                Meta :: rabbit_fifo:consumer_meta(),
-               state()) -> {ok, state()} | {error | timeout, term()}.
-checkout(ConsumerTag, NumUnsettled, CreditMode, Meta,
-         #state{consumer_deliveries = CDels0} = State0) ->
+               state()) ->
+    {ok, ConsumerInfos :: map(), state()} |
+    {error | timeout, term()}.
+checkout(ConsumerTag, CreditMode, #{} = Meta,
+         #state{consumers = CDels0} = State0)
+  when is_binary(ConsumerTag) andalso
+       is_tuple(CreditMode) ->
     Servers = sorted_servers(State0),
-    ConsumerId = {ConsumerTag, self()},
-    Cmd = rabbit_fifo:make_checkout(ConsumerId,
-                                    {auto, NumUnsettled, CreditMode},
-                                    Meta),
+    ConsumerId = consumer_id(ConsumerTag),
+    Spec = case rabbit_fifo:is_v4() of
+               true ->
+                   case CreditMode of
+                       {simple_prefetch, 0} ->
+                           {auto, {simple_prefetch,
+                                   ?UNLIMITED_PREFETCH_COUNT}};
+                       _ ->
+                           {auto, CreditMode}
+                   end;
+               false ->
+                   case CreditMode of
+                       {credited, _} ->
+                           {auto, 0, credited};
```

Comment:
When I enable feature flag `credit_api_v2` but leave feature flag `quorum_queues_v2` disabled and attach an AMQP 1.0 receiver to a quorum queue, the session proc crashes:
```
2024-07-17 12:42:15.626243+00:00 [error] <0.705.0>   crasher:
2024-07-17 12:42:15.626243+00:00 [error] <0.705.0>     initial call: rabbit_amqp_session:init/1
2024-07-17 12:42:15.626243+00:00 [error] <0.705.0>     pid: <0.705.0>
2024-07-17 12:42:15.626243+00:00 [error] <0.705.0>     registered_name: []
2024-07-17 12:42:15.626243+00:00 [error] <0.705.0>     exception error: no function clause matching
2024-07-17 12:42:15.626243+00:00 [error] <0.705.0>                      rabbit_amqp_session:'-handle_queue_actions/2-fun-1-'({send_credit_reply,
2024-07-17 12:42:15.626243+00:00 [error] <0.705.0>                                                                            0},
2024-07-17 12:42:15.626243+00:00 [error] <0.705.0>                                                                           {state,
...
(rabbit_amqp_session.erl, line 1940)
2024-07-17 12:42:15.626243+00:00 [error] <0.705.0>       in function  lists:foldl/3 (lists.erl, line 1594)
2024-07-17 12:42:15.626243+00:00 [error] <0.705.0>       in call from rabbit_amqp_session:handle_cast/2 (rabbit_amqp_session.erl, line 476)
2024-07-17 12:42:15.626243+00:00 [error] <0.705.0>       in call from gen_server:try_handle_cast/3 (gen_server.erl, line 1121)
2024-07-17 12:42:15.626243+00:00 [error] <0.705.0>       in call from gen_server:handle_msg/6 (gen_server.erl, line 1183)
```

This is because the session thinks it uses credit API v2 although quorum queue client and quorum queue server decide to use credit API v1 under the hood.

To solve this, we could either check whether `quorum_queues_v4` is enabled in the session proc and only then use credit API v2.
An alternative solution is to merge both feature flags `credit_api_v2` and `quorum_queues_v4` into a single feature flag `queue_api_4_0` which includes all new classic queue client to classic queue server and quorum queue client to quorum queue server interactions introduced in RabbitMQ 4.0
(Feature flag `credit_api_v2` is arguably misnamed because it includes already both credit API v2 and cancel API v2.)

Reply:
Fixed by merging all feature flags introduced in RabbitMQ 4.0.0 into a single feature flag `rabbitmq_4.0.0`.